### ISSUE TITLE
Bars added by other mods are ignored

### ISSUE THREAD

**Issue by noawx** [MDLC01/higher-chat-mc#4]
Version: Fabric 1.20.1
Higher Chat Version: 1
[Armor Points ++](https://modrinth.com/mod/armorpoints) Version: fabric-1.20-3.1.1

When using Armor Points ++ and Higher Chat, the chat returns to it's original position.

![image](https://github.com/MDLC01/higher-chat-mc/assets/76262280/08a910fb-f771-445d-a1bd-627ea3d2e873)

**Comment by MDLC01** [comment]
As explained in the [README](https://github.com/MDLC01/higher-chat-mc#is-it-compatible-with-mods-that-add-other-bars), only vanilla bars are taken into account by Higher Chat. Sadly, it looks like Armor Points ++ reimplements those bars instead of modifying them.

Making Higher Chat work with other mods has to be done for each mod individually. I might look into making it compatible with Armor Points ++ in the future, but do not expect a fix soon.

I'm leaving the issue open for now, as I won't go as far as saying it is not planned.

**Comment by noawx** [comment]
Maybe there could be a config to switch between automatically determining the position of the chat and being able to manually set a static position? (for example: you can make the chat always stay 2 bars higher)

**Comment by MDLC01** [comment]
That could be an option indeed, although that would require the Fabric API and/or [Mod Menu](https://modrinth.com/mod/modmenu).

I will try to implement it with minimal requirements when I get the time.

**Comment by Cheos137** [comment]
Hi there, author of Armorpoints++ here

looking at this, there are three solutions i can think of:
1. Armorpoints++ calling `SharedStorage.declareIconAt(x, y);` whenever it detects higher-chat being loaded
2. higher-chat checking the `AppGui#left_offset` and `AppGui#right_offset` fields when it detects armorpointspp being loaded
3. higher-chat supporting [libhud](https://modrinth.com/mod/libhud) (a mod i created to hopefully finally clear this mess of hud rendering incompatibilities at some point in the future) and checking heights in a similar manner as in 2 or by listening to component render events ([libhud docs](https://github.com/Cheos137/libhud-fabric/wiki/Developing-with-libhud))

3 is a bit special because it kind of exists in parallel to the other two options.

personally i'd prefer for [libhud](https://modrinth.com/mod/libhud) support being pushed, however i will try to assist you in making higher-chat and Armorpoints++ compatible either way

Edit: added links + the following note
Note: both Armorpoints++ and libhud inject their respective gui instance into the minecraft client in a way that you can obtain it via the `Minecraft.getInstance().gui` field and then casting it to the respective gui class

**Comment by MDLC01** [comment]
Option 3 seems to be the most reasonable to me.

I don't like option 1 because this "declaring icons" system is already what causes #2, so I wanted to get rid of it anyway; and option 2 is not ideal because it only fixes the problem with Armor Points ++ instead of a more general solution for all mods that modify or add bars.

My only concern with option 3 is that it requires libhud to become a dependency of Higher Chat, which is annoying for users who don't need it.

**Comment by Cheos137** [comment]
you could keep the current system to be used as a fallback (even if it isn't perfect) when libhud is not installed so that it would work for everyone out of the box

**Comment by MDLC01** [comment]
I have created a `libhud` branch on this repo that contains changes that should make Higher Chat compatible with libhud. Do you see any potential problem with this implementation (I have already tested it both with and without libhud)?

There are currently two differences in behavior:
1. The chat is raised even when the armor bar is not present with libhud, because `leftOffset` is always incremented in [`dev.cheos.libhud.LibhudGui#renderArmor`](https://github.com/Cheos137/libhud-fabric/blob/b8549df799414f67e6bdefaff79a0d7ade8cc84f/src/main/java/dev/cheos/libhud/LibhudGui.java#L208-L232);
2. #2 is fixed when libhud is present.

If this first implementation is fine, I will clean it up and build a new version of Higher Chat with it as soon as possible.

**Comment by Cheos137** [comment]
your implementation looks good to me ^^


> The chat is raised even when the armor bar is not present with libhud, because leftOffset is always incremented in [dev.cheos.libhud.LibhudGui#renderArmor](https://github.com/Cheos137/libhud-fabric/blob/b8549df799414f67e6bdefaff79a0d7ade8cc84f/src/main/java/dev/cheos/libhud/LibhudGui.java#L208-L232);

Whoops, gonna fix that

EDIT: fix build is up on modrinth: https://modrinth.com/mod/libhud/version/1.0.0-beta3

**Comment by MDLC01** [comment]
`main` is now fully compatible with libhud.

I will be working on supporting Armor Points ++ as well soon.

**Comment by Cheos137** [comment]
unsure whether dedicated Armorpoints++ support is really needed, as it fully supports libhud, too

gonna put some time into moving the compat modules of Armorpoints++ into a dedicated (not officially supported) helper mod so that i can then start pushing it and get more modders to either support it or fully build onto it

however, if you want to add dedicated Armorpoints++ support, i won't stop you :)

**Comment by MDLC01** [comment]
> unsure whether dedicated Armorpoints++ support is really needed, as it fully supports libhud, too

Well... Higher Chat currently does not work with Armor Points ++, just like before libhud support.

> gonna put some time into moving the compat modules of Armorpoints++ into a dedicated (not officially supported) helper mod so that i can then start pushing it and get more modders to either support it or fully build onto it

Sorry, I'm not sure what you mean by that. Will Armor Points ++ become compatible with Higher Chat with this change?

**Comment by Cheos137** [comment]
Wait, Armorpoints++ breaks higher chat even with libhud installed?

**Comment by MDLC01** [comment]
Yes, because `Minecraft.getInstance().gui` is not a `LibhudGui`, so the cast does not happen here:
https://github.com/MDLC01/higher-chat-mc/blob/1dfd14ef8b3f42f7edbb39f592ffc345e08b9a09/src/main/java/com/mdlc/higherchat/SharedStorage.java#L68

**Comment by Cheos137** [comment]
yeah, but that should only be the case if libhud is not installed... unless i messed something up badly

**Comment by MDLC01** [comment]
Mmh that's weird. I remember noticing this bug, but I apparently did not think this was a problem at the time for some reason.

![The hearts nor the armor bar are visible. Instead, there are black squares.](https://github.com/MDLC01/higher-chat-mc/assets/57839069/b5ed0665-7c2d-4235-9321-a78717b08a1e)


I should have 40 HP (so two full rows of hearts).

Am I missing something obvious?

I get the following error:

```
Failed to load texture: armorpointspp:textures/gui/icons.png
java.io.FileNotFoundException: armorpointspp:textures/gui/icons.png
	at net.minecraft.class_5912.method_43043(class_5912.java:21)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at net.minecraft.class_5912.getResourceOrThrow(class_5912.java:21)
	at net.minecraft.class_1049$class_4006.method_18156(class_1049.java:83)
	at net.minecraft.class_1049.method_18153(class_1049.java:58)
	at net.minecraft.class_1049.method_4625(class_1049.java:29)
	at net.minecraft.class_1060.method_24303(class_1060.java:97)
	at net.minecraft.class_1060.method_4616(class_1060.java:69)
	at net.minecraft.class_1060.method_4619(class_1060.java:119)
	at com.mojang.blaze3d.systems.RenderSystem._setShaderTexture(RenderSystem.java:1232)
	at com.mojang.blaze3d.systems.RenderSystem.setShaderTexture(RenderSystem.java:1225)
	at net.minecraft.class_332.method_25295(class_332.java:387)
	at net.minecraft.class_332.method_25297(class_332.java:378)
	at net.minecraft.class_332.method_25293(class_332.java:362)
	at net.minecraft.class_332.method_25290(class_332.java:374)
	at dev.cheos.armorpointspp.impl.RendererImpl.blit(RendererImpl.java:30)
	at dev.cheos.armorpointspp.core.texture.ITextureSheet.blit(ITextureSheet.java:36)
	at dev.cheos.armorpointspp.core.texture.StandardTextureSheet.drawHeartBG(StandardTextureSheet.java:46)
	at dev.cheos.armorpointspp.core.render.HealthComponent.render(HealthComponent.java:71)
	at dev.cheos.armorpointspp.Overlays.playerHealth(Overlays.java:32)
	at dev.cheos.armorpointspp.ApppGui.method_1760(ApppGui.java:54)
	at net.minecraft.class_329.method_1753(class_329.java:214)
	at dev.cheos.armorpointspp.ApppGui.method_1753(ApppGui.java:41)
	at net.minecraft.class_757.method_3192(class_757.java:926)
	at net.minecraft.class_310.method_1523(class_310.java:1219)
	at net.minecraft.class_310.method_1514(class_310.java:802)
	at net.minecraft.client.main.Main.main(Main.java:250)
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:468)
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74)
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23)
```

And the mod list:

```
Loading 7 mods:
	- armorpointspp 3.1.1
	   \-- fiber 0.23.0-2
	- fabricloader 0.14.21
	- higher-chat dev
	- java 17
	- libhud 1.20-1.0.0-beta4
	- minecraft 1.20.1
```

**Comment by Cheos137** [comment]
that's really weird.. gonna look into it tomorrow

**Comment by Cheos137** [comment]
ok, so...
Armorpoints++ ignoring libhud being present: fixed in 3.1.2 (up soon) / libhud-1.0.0-beta6 for some additional fixes
and the texture issue seems to be that mod resources only get loaded if fabric api is present? (which i don't recall happening previously)

**Comment by Cheos137** [comment]
(also tested with higher chat and can confirm compatibility working with libhud)
